src/cluster_legacy.c

```diff
@@ -3638,16 +3639,22 @@ int clusterProcessPacket(clusterLink *link) {
                              * this may cause a primary-replica chain issue. */
                             return 1;
                         } else if (nodeIsReplica(sender_claimed_primary)) {
+                            serverAssert(sender_claimed_primary->replicaof == sender);
```

Comment:
I don't see why this must be true at this point in time in the code. It also is crashing here: https://github.com/valkey-io/valkey/actions/runs/16720515261/job/47323708440?pr=2421#step:5:6544.

Reply:
The CI did not print the server logs, so i'm still guessing what the scenario might be.

i add this assert because, `sender_claimed_primary` and `sender` are in the same shard, and `sender` is the old primary, so `sender` should be the primary of `sender_claimed_primary`

Reply:
i guess the scenario could be:
- sender is an empty primary
- sender_claimed_primary doing a failover in its shard
- sender replicate to sender_claimed_primary

Reply:
i think the problem could be #2279, like due to some reasons, there are two primaries in the same shard

Reply:
ok, i combined the tests in #2279 (two primaries in the same shard) and #2301 (this one) and successfully reproduced this assert.

so there will be a case that can reproduce this assert:
1. pre condition: two primaries in the same shard (from #2279)
2. replica doing the failover
3. the empty primary doing a replicate
4. they are in the same shard and meet every conditions in the code and trigger the assert

Reply:
Great! Now, to keep the two PRs separate, it's safer to just remove the assert, or change it to an `if` condition?

```c
                        } else if (nodeIsReplica(sender_claimed_primary) &&
                                   sender_claimed_primary->replicaof == sender)) {
```

Reply:
> Great! Maybe it's safer to just remove the assert, or change it to an `if` condition?
> 
> ```c
>                         } else if (nodeIsReplica(sender_claimed_primary) &&
>                                    sender_claimed_primary->replicaof == sender)) {
> ```